### ISSUE TITLE
unnecessary_to_owned suggests doing more allocation

### ISSUE THREAD

**Issue by ndmitchell** [rust-lang/rust-clippy#8914]
### Summary

The suggestion is significantly less efficient than the code it replaces, requiring _O(n)_ `clone`s.

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

```rust
fn foo(x: Box<[String]>) -> impl Iterator<Item = String> {
    x.to_vec().into_iter()
}
```

I saw this happen:

```
warning: unnecessary use of `to_vec`
 --> src/lib.rs:3:5
  |
3 |     x.to_vec().into_iter()
  |     ^^^^^^^^^^^^^^^^^^^^^^ help: use: `x.iter().cloned()`
  |
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
```

I expected to see this happen:

Nothing - the original is efficient non-cloning code. The suggested version clones each string and is thus significantly less efficent.

### Version

```text
0.1.63 (2022-05-29 28b8919) (Ran on Rust Playground)
```

**Comment by Alexendoo** [comment]
`.to_vec` does clone the contents of the slice, perhaps you meant to use `.into_vec` instead

**Comment by ndmitchell** [comment]
Thanks, that is indeed the case, sorry for the noise. I wonder if `x.to_vec()` where `x` is a `Box<[T]>` you have ownership of should be a `into_vec()`, but that seems a lot harder to figure out.
